planner/core/rule_generate_column_substitute.go

```diff
@@ -66,7 +66,7 @@ func collectGenerateColumn(lp LogicalPlan, exprToColumn ExprColumnMap) {
 			if colInfo.IsGenerated() && !colInfo.GeneratedStored {
 				s := ds.schema.Columns
 				col := expression.ColInfo2Col(s, colInfo)
-				if col != nil && col.GetType().Equal(col.VirtualExpr.GetType()) {
+				if col != nil && col.GetType().PartialEqual(col.VirtualExpr.GetType(), lp.SCtx().GetSessionVars().EnableUnsafeSubstitute) {
```

Comment:
Maybe we can add an example to explain why we want to enable unsafe substitution to make it easy to understand.

Reply:
done, add an issue link